### ISSUE TITLE
simple_query_string returns nothing for "*" search

### ISSUE THREAD

**Issue by pierrre** [elastic/elasticsearch#10632]
My simple_query_string returns nothing with "*" search.

If I use a query_string, it works (it returns hits).
The fields and analyzer are the same.

If I wrap my simple_query_string with a bool+must (or should) query, it works. (???)

Is it a bug?

Elasticsearch version: 1.5.1

**Comment by dakrone** [comment]
I believe this may be related to #9866, which was fixed in Lucene 5.1 (see: https://issues.apache.org/jira/browse/LUCENE-6298). I'll look and see whether it is the same issue.

**Comment by pierrre** [comment]
My Lucene version is 4.10.4.

What should happen if I search "*" with a simple_query_string?
Return all docs or nothing?
The query_string returns all docs.

**Comment by clintongormley** [comment]
This has been fixed.  * will return all docs

**Comment by pierrre** [comment]
Thank you!

**Comment by pierrre** [comment]
Actually, this search:
```
{
  "fields": [],
  "query": {
    "simple_query_string": {
      "fields": [
        "shortDescription"
      ],
      "query": "*"
    }
  }
}
```
returns nothing with Elasticsearch 2.2

@clintongormley is the fix for this issue released?

**Comment by clintongormley** [comment]
Initially I thought this might be related to https://github.com/elastic/elasticsearch/issues/13214 but no, a query string of `"*"` doesn't match any docs, with or without custom `fields`:

    PUT t/t/1
    {}
    
    GET t/_search
    {
      "query": {
        "simple_query_string": {
          "query": "*"
        }
      }
    }
    
    GET _validate/query?explain
    {
      "query": {
        "simple_query_string": {
          "query": "*"
        }
      }
    }

The explanation returned by the last request is simply `""`

**Comment by cbuescher** [comment]
I looked into this a bit by adding an IT test. It looks like Lucenes SimpleQueryParser only recognizes a terms with a closing '*' as prefix query. A single wildcard without leading charactes is treated as a regular term in SimpleQueryParser#consumeToken(). This might even be the intended behaviour. For `query_string`  the situation is different since it uses MapperQueryParser internally which has an `allowLeadingWildcard` option. Maybe @jimferenczi or @dakrone have an opinion on whether this can or should be fixed in SimpleQueryParser or is even intended.

**Comment by dakrone** [comment]
@cbuescher personally I think a SQS query for `*` should be internally rewritten to a `MatchAllDocsQuery`, which I think means this would go into Lucene. If people agree I can work on a patch for the next Lucene release

**Comment by clintongormley** [comment]
@dakrone not so sure - should it be match all or should it match all docs that have a value for the specified fields?  See #13214 for a similar discussion

**Comment by dakrone** [comment]
@clintongormley ahh okay, in light of that discussion:

>    "*" always means match all documents
 >   "field:*" means find all documents that contain a non-null value in the field (incl empty string)

Since `simple_query_string` can either use the `default_field` or have `fields`
specified, I'm thinking we should do:

- If query is `"*"` and no `fields` are specified, treat it as match all
  documents
- If query is `"*"` and `fields` are specified, query is essentially `field1:*
  OR field2:* OR field3:*` which matches documents that have non-null values in
  the field

Do you think that'd be a better solution?

**Comment by clintongormley** [comment]
@dakrone Given that the end user can't specify fields in the SQS like you can in QS (`user:kimchy`) I've come around to thinking that `*` should always mean match all docs.

**Comment by dakrone** [comment]
I opened https://issues.apache.org/jira/browse/LUCENE-7490 for this with a patch

**Comment by dakrone** [comment]
Looks like this will be in Lucene 6.3, so I'm marking this as "stalled" until 6.3 is released and incorporated.

**Comment by weberhofer** [comment]
I think this issue has been solved with ES 5.3

**Comment by cbuescher** [comment]
@weberhofer thanks a lot, I just checked with 5.1 (which is based in Lucene 6.3) and 5.2 and it seems to be fixed. I will close this.
